Generate generics instead of decoders with macros

emdash-ie opened the pull request:
This PR replaces the macros which generated circe Encoder and Decoders for scrooge-generated classes for thrift structs and unions with new macros which instead generate shapeless LabelledGeneric values. These values match the values that shapeless would generate for the equivalent case classes (mostly; there are some changes to match the existing behaviour with default values), which allows us to use circe’s semi-auto derivation to get the encoders and decoders.

This has the advantage that we no longer need to maintain as much macro complexity, and we can now create encoders/decoders for recursive struct hierarchies (either direct or indirect).

Hopefully this will also mean we can migrate to scala 3 more easily, as our macros are less complicated and more connected to shapeless, which we hear scala 3 macros are inspired by.

Unfortunately, circe’s fully automatic derivation no longer works with this approach, which means this is a breaking change for consumers of the library.

## Implementation notes

- To get it working I had to use whitebox macros instead of blackbox macros, but unfortunately the existing macros don’t work when they’re whitebox, so I’ve split the new macros into their own file (I can’t see how to have whitebox and blackbox macros in the same file).

  - Whitebox implicit def macros [allow refining of the result type by the macro](https://docs.scala-lang.org/overviews/macros/blackbox-whitebox.html). I.e., the implicit search can begin for a `LabelledGeneric.Aux[Union1, R]` where `R` is completely unconstrained, and once the macro has run the result can be examined to refine this type to `LabelledGeneric.Aux[Union1, Union1Repr]`, where `Union1Repr` is a specific HList type (which implicit search can then find a circe macro for to generate a decoder). 

- To match the existing decoder behaviour, any struct field with a default parameter is allowed to be missing from the json, in which case the default value will be used.
  
- Fezziwig now serialises missing optional values as explicit nulls, rather than missing fields. This changes the contract with consumers of the library, but consumers can use deepDropNullValues to exclude nulls, which matches how circe is used elsewhere.

- I suspect there’s a way to replace these macros altogether with shapeless-style implicits, but I don’t know how yet and this works.


## How to test

This change is pretty thoroughly tested by the repository tests, as we’ve added some while it was worked on.

## How can we measure success?

- Recursive structs can now be used with the library
- Hopefully, migration to scala 3 in the future will be easier

## Have we considered potential risks?

### Semi-auto derivation

I believe the main risk is that semi-auto derivation will be more of a pain for consumers of the library than we anticipate.

I’ve made the necessary changes to content-api-models to work with this PR in [content-api-models PR 239](https://github.com/guardian/content-api-models/pull/239), and I don’t think it’s _too_ bad.

### Compile times

It’s hard to analyse the effect these changes will have on compile times: the generated macros are much smaller for nested structs, but more implicit searching is required, and I think whitebox macros are probably less performant than blackbox macros?

I’ll try to make some measurements for content-api-models!

### Non-isomorphism

I think treating default values the way we do also means that the generated LabelledGeneric is not quite an isomorphism: if you start with a thrift struct with an optional field which has a default value, and convert an instance of that struct where the field’s value is unspecified to the generic representation and back, you will now have a thrift struct where the value is specified and matches the default value.

This may cause undesirable or confusing behaviour if using the generic instance outside of json, for example if using shapeless’ lenses or zippers.

emdash-ie committed:
Generate generics with macros instead of decoders

This commit replaces the macros which generated circe Encoder and
Decoders for scrooge-generated classes for thrift structs and unions
with new macros which instead generate shapeless LabelledGeneric values.
These values match the values that shapeless would generate
(mostly; there are some changes to match the existing behaviour with
default values) for the equivalent case classes, which allows us to use
circe’s semi-auto derivation to get the encoders and decoders.

This has the advantage that we no longer need to maintain as much macro
complexity, and we can now create encoders/decoders for recursive struct
hierarchies (either direct or indirect).

Hopefully this will also mean we can migrate to scala 3 more easily, as
our macros are less complicated and more connected to shapeless, which
we hear scala 3 macros are inspired by.

Unfortunately, circe’s fully automatic derivation no longer works with
this approach, which means this is a (small, I think) breaking change
for consumers of the library.

Implementation notes:

- To get it working I had to use whitebox macros instead of blackbox
  macros, but unfortunately the existing macros don’t work when they’re
  whitebox, so I’ve split the new macros into their own file (I can’t
  see how to have whitebox and blackbox macros in the same file).

- This method no longer suppresses nulls in the output, but I believe
  that’s possible for consumers using a circe option.

- I suspect there’s a way to replace these macros altogether with
  shapeless-style implicits, but I don’t know how yet and this
  works.

- To match the existing decoder behaviour, any struct field with a
  default parameter is allowed to be missing from the json, in which
  case the default value will be used.

davidfurey committed:
Factor out testRoundTrip

davidfurey committed:
Add recursive tests based on Apache Thrift project

See https://github.com/apache/thrift/blob/master/test/Recursive.thrift

emdash-ie commented on src/main/scala/com/gu/fezziwig/CirceScroogeMacros.scala:
```diff
@@ -4,7 +4,12 @@ import scala.language.experimental.macros
+import shapeless.{Lazy, |¬|, LabelledGeneric}
+import shapeless.HList
+
+trait Representation[A] {
+  type R
+}
```
This type should be removed: I didn’t end up using it for anything

davidfurey committed:
Add test for desired new CAPI model

davidfurey committed:
Add failing test for optional fields

davidfurey committed:
Fix null values test

Use deepDropNullValues to exclude nulls rather than handling this in fezziwig.

NB- changes the contract with consumers of the library but matches how circe is used elsewhere.

emdash-ie commented:
I’ve made some headway on default values, but am not quite there yet: 

- for default values where the type in the hlist is currently an Option, we can change the `from` method in the generic implicit to add `.orElse(<default-value>)`
- however, scrooge doesn’t use an Option for optional fields which have a default value specified (e.g. `WithDefault.something`)
- also, scrooge provides a default value for required fields whose type is a map, e.g. in `StructA.apply`, the default value for the `intMap` parameter is `Map.empty`

I think using an `Option` in the generic representation for any parameter with a default value will work (always wrapping in `Some` when converting to the repr type and replacing `None` with the default value when converting back), but I haven’t fully thought it through yet so there may be further complication there. (E.g. will that allow parameters that are supposed to be required to be left empty, just because scrooge uses a default value?)

emdash-ie commented on src/main/scala/com/gu/fezziwig/CirceScroogeWhiteboxMacros.scala:
```diff
@@ -0,0 +1,301 @@
+    *     def to(union: Union1): Repr = union match {
+    *       case Union1.D(x) => Inl(labelled.field(x))
+    *       case Union1.C(x) => Inr(Inl(labelled.field(x)))
+    *       case Union1.UnknownUnionField(_) => throw new RuntimeException("Unknown union field")
```
Should this handle unknown union fields better? How is this handled in the current implementation? I will check.

emdash-ie replied:
Given the way we serialise I’m not certain how we would detect unknown union fields: I suppose we’d look for any keys that exist on the object and read the first one we find as the unknown union field?

emdash-ie replied:
In the current implementation, the encoder serialises `UnknownUnionField` values as `null`. The decoder doesn’t ever produce a value of type `UnknownUnionField`, and returns a `DecodingFailure` for `null`.

emdash-ie replied:
I also found a bug in the decoder macro, I think: it calls `Either.left` in it, which doesn’t exist. I’m pretty sure it’s meant to be `_root_.scala.util.Left` instead. I don’t understand why that only fails to typecheck if I explicitly materialise the decoder though – our tests successfully decode values of union types in fields of structs, which is a bit mysterious.

emdash-ie replied:
I don’t think I can easily match the current behaviour: I’m not sure how to create a generic representation whose derived decoder will fail to decode what the derived encoder writes. But I don’t think having a non-roundtripping encoder/decoder pair is a good idea anyway.

So I think I should consider incorporating unknown fields into the representation. I guess this would be an extra `:+:` branch, either using a field name that’s likely to be uncommon or automatically picking a field name that’s unused by the union. I think the former is probably fine, as long as I pick a very awkward name – thinking about how to deterministically choose an unused name (so that it’s the same every time the generic is derived) makes me uneasy.

emdash-ie replied:
I have now updated this PR to handle unknown fields by preserving the field names, types, and ids, but dropping the field values.

emdash-ie commented on src/main/scala/com/gu/fezziwig/CirceScroogeWhiteboxMacros.scala:
```diff
@@ -0,0 +1,301 @@
+    *     def from(repr: Repr): Union1 = repr match {
+    *       case Inl(d) => Union1.D(d)
+    *       case Inr(Inl(c)) => Union1.C(c)
+    *       case Inr(Inr((_: CNil))) => throw new RuntimeException("Encountered CNil while converting from ReprType")
```
Note that this case can’t happen, because it’s not possible to construct a value of type CNil. The compiler successfully detects this and warns here, but it also warns about an incomplete match if I leave the case out. I guess I figured the explicit error is better than an implicit one?

emdash-ie replied:
I wonder if an incomplete match would be better.

emdash-ie replied:
I’ve realised that taking out the `: CNil` type signature removes the warning, so I’m gonna go with that to avoid putting annoying warnings in consumers’ code.

emdash-ie committed:
Remove InnerStruct and OuterStruct test

This doesn’t test anything now that’s not covered by another test.

emdash-ie commented:
Note that this PR targets the branch for [pull 49](https://github.com/guardian/fezziwig/pull/49), as I added some tests there and then copied them onto this branch.

davidfurey commented:
Are these two points in the description the same?

> This method no longer suppresses nulls in the output, but I believe that’s possible for consumers using a circe option.

> Fezziwig now serialises missing optional values as explicit nulls, rather than missing fields. This changes the contract with consumers of the library, but consumers can use deepDropNullValues to exclude nulls, which matches how circe is used elsewhere.

emdash-ie commented:
> Are these two points in the description the same?
> 
> > This method no longer suppresses nulls in the output, but I believe that’s possible for consumers using a circe option.
> 
> > Fezziwig now serialises missing optional values as explicit nulls, rather than missing fields. This changes the contract with consumers of the library, but consumers can use deepDropNullValues to exclude nulls, which matches how circe is used elsewhere.

Yes, I forgot to remove my note when adding/adapting yours (which had more detail). I’ll remove the first of the two now.

emdash-ie committed:
Remove necessity for manual semi-auto derivation

Adding these implicits has fixed the problem, and implicit search can
now find the decoders correctly when they’re needed. I’m not sure
exactly why this is necessary, but I also don’t really understand how
circe’s auto-derivation works.

emdash-ie commented:
The latest commit removes the necessity for semi-auto derivation, which should make this less painful to use.

emdash-ie commented:
It seems this approach causes compilation times to slow way down. I don’t know how to fix it yet, but I think I know the cause: modifying the `thriftStructGeneric` macro to print which types it’s being run for each time and running the tests shows that the macro is being called for many types that are not thrift structs or unions, and repeatedly for the same types. 

I think shapeless’ `cachedImplicit` could help with the second problem, but what about the first? It surprises me greatly that the macro is being called for non-struct types (e.g. `Option[Int]`!), given we have an upper bound in the type parameter definition.

<details>
<summary>Expand for list of types</summary>
```
Making a LabelledGeneric[com.gu.fezziwig.RecTree]
Making a LabelledGeneric[Option[scala.collection.Seq[com.gu.fezziwig.RecTree]]]
Making a LabelledGeneric[Some[scala.collection.Seq[com.gu.fezziwig.RecTree]]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecTree]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecTree]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecTree]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecTree]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecTree]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecTree]]
Making a LabelledGeneric[com.gu.fezziwig.RecTree]
Making a LabelledGeneric[Option[scala.collection.Seq[com.gu.fezziwig.RecTree]]]
Making a LabelledGeneric[Some[scala.collection.Seq[com.gu.fezziwig.RecTree]]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecTree]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecTree]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecTree]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecTree]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecTree]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecTree]]
Making a LabelledGeneric[com.gu.fezziwig.RecTree]
Making a LabelledGeneric[Option[scala.collection.Seq[com.gu.fezziwig.RecTree]]]
Making a LabelledGeneric[Some[scala.collection.Seq[com.gu.fezziwig.RecTree]]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecTree]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecTree]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecTree]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecTree]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecTree]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecTree]]
Making a LabelledGeneric[com.gu.fezziwig.RecTree]
Making a LabelledGeneric[Option[scala.collection.Seq[com.gu.fezziwig.RecTree]]]
Making a LabelledGeneric[Some[scala.collection.Seq[com.gu.fezziwig.RecTree]]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecTree]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecTree]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecTree]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecTree]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecTree]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecTree]]
Making a LabelledGeneric[com.gu.fezziwig.RecList]
Making a LabelledGeneric[Option[Option[com.gu.fezziwig.RecList]]]
Making a LabelledGeneric[Some[Option[com.gu.fezziwig.RecList]]]
Making a LabelledGeneric[Option[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[Some[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[com.gu.fezziwig.RecList]
Making a LabelledGeneric[Option[Option[com.gu.fezziwig.RecList]]]
Making a LabelledGeneric[Some[Option[com.gu.fezziwig.RecList]]]
Making a LabelledGeneric[Option[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[Some[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[com.gu.fezziwig.RecList]
Making a LabelledGeneric[Option[Option[com.gu.fezziwig.RecList]]]
Making a LabelledGeneric[Some[Option[com.gu.fezziwig.RecList]]]
Making a LabelledGeneric[Option[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[Some[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[com.gu.fezziwig.RecList]
Making a LabelledGeneric[Option[Option[com.gu.fezziwig.RecList]]]
Making a LabelledGeneric[Some[Option[com.gu.fezziwig.RecList]]]
Making a LabelledGeneric[Option[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[Some[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[com.gu.fezziwig.CoRec]
Making a LabelledGeneric[com.gu.fezziwig.CoRec2]
Making a LabelledGeneric[Option[Option[com.gu.fezziwig.CoRec]]]
Making a LabelledGeneric[Some[Option[com.gu.fezziwig.CoRec]]]
Making a LabelledGeneric[Option[com.gu.fezziwig.CoRec]]
Making a LabelledGeneric[Some[com.gu.fezziwig.CoRec]]
Making a LabelledGeneric[com.gu.fezziwig.CoRec]
Making a LabelledGeneric[com.gu.fezziwig.CoRec2]
Making a LabelledGeneric[Option[Option[com.gu.fezziwig.CoRec]]]
Making a LabelledGeneric[Some[Option[com.gu.fezziwig.CoRec]]]
Making a LabelledGeneric[Option[com.gu.fezziwig.CoRec]]
Making a LabelledGeneric[Some[com.gu.fezziwig.CoRec]]
Making a LabelledGeneric[com.gu.fezziwig.CoRec]
Making a LabelledGeneric[com.gu.fezziwig.CoRec2]
Making a LabelledGeneric[Option[Option[com.gu.fezziwig.CoRec]]]
Making a LabelledGeneric[Some[Option[com.gu.fezziwig.CoRec]]]
Making a LabelledGeneric[Option[com.gu.fezziwig.CoRec]]
Making a LabelledGeneric[Some[com.gu.fezziwig.CoRec]]
Making a LabelledGeneric[com.gu.fezziwig.CoRec]
Making a LabelledGeneric[com.gu.fezziwig.CoRec2]
Making a LabelledGeneric[Option[Option[com.gu.fezziwig.CoRec]]]
Making a LabelledGeneric[Some[Option[com.gu.fezziwig.CoRec]]]
Making a LabelledGeneric[Option[com.gu.fezziwig.CoRec]]
Making a LabelledGeneric[Some[com.gu.fezziwig.CoRec]]
Making a LabelledGeneric[com.gu.fezziwig.VectorTest]
Making a LabelledGeneric[Option[scala.collection.Seq[com.gu.fezziwig.RecList]]]
Making a LabelledGeneric[Some[scala.collection.Seq[com.gu.fezziwig.RecList]]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[com.gu.fezziwig.RecList]
Making a LabelledGeneric[Option[Option[com.gu.fezziwig.RecList]]]
Making a LabelledGeneric[Some[Option[com.gu.fezziwig.RecList]]]
Making a LabelledGeneric[Option[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[Some[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[com.gu.fezziwig.VectorTest]
Making a LabelledGeneric[Option[scala.collection.Seq[com.gu.fezziwig.RecList]]]
Making a LabelledGeneric[Some[scala.collection.Seq[com.gu.fezziwig.RecList]]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[com.gu.fezziwig.RecList]
Making a LabelledGeneric[Option[Option[com.gu.fezziwig.RecList]]]
Making a LabelledGeneric[Some[Option[com.gu.fezziwig.RecList]]]
Making a LabelledGeneric[Option[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[Some[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[com.gu.fezziwig.VectorTest]
Making a LabelledGeneric[Option[scala.collection.Seq[com.gu.fezziwig.RecList]]]
Making a LabelledGeneric[Some[scala.collection.Seq[com.gu.fezziwig.RecList]]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[com.gu.fezziwig.RecList]
Making a LabelledGeneric[Option[Option[com.gu.fezziwig.RecList]]]
Making a LabelledGeneric[Some[Option[com.gu.fezziwig.RecList]]]
Making a LabelledGeneric[Option[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[Some[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[com.gu.fezziwig.VectorTest]
Making a LabelledGeneric[Option[scala.collection.Seq[com.gu.fezziwig.RecList]]]
Making a LabelledGeneric[Some[scala.collection.Seq[com.gu.fezziwig.RecList]]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[com.gu.fezziwig.RecList]
Making a LabelledGeneric[Option[Option[com.gu.fezziwig.RecList]]]
Making a LabelledGeneric[Some[Option[com.gu.fezziwig.RecList]]]
Making a LabelledGeneric[Option[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[Some[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.RecList]]
Making a LabelledGeneric[com.gu.fezziwig.BlockElement]
Making a LabelledGeneric[Option[Option[com.gu.fezziwig.ListElementFields]]]
Making a LabelledGeneric[Some[Option[com.gu.fezziwig.ListElementFields]]]
Making a LabelledGeneric[Option[com.gu.fezziwig.ListElementFields]]
Making a LabelledGeneric[Some[com.gu.fezziwig.ListElementFields]]
Making a LabelledGeneric[com.gu.fezziwig.ListElementFields]
Making a LabelledGeneric[Option[scala.collection.Seq[com.gu.fezziwig.ListItem]]]
Making a LabelledGeneric[Some[scala.collection.Seq[com.gu.fezziwig.ListItem]]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.ListItem]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.ListItem]]
Making a LabelledGeneric[com.gu.fezziwig.ListItem]
Making a LabelledGeneric[Option[scala.collection.Seq[com.gu.fezziwig.BlockElement]]]
Making a LabelledGeneric[Some[scala.collection.Seq[com.gu.fezziwig.BlockElement]]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.BlockElement]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.BlockElement]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.BlockElement]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.BlockElement]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.BlockElement]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.BlockElement]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.ListItem]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.ListItem]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.ListItem]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.ListItem]]
Making a LabelledGeneric[com.gu.fezziwig.BlockElement]
Making a LabelledGeneric[Option[Option[com.gu.fezziwig.ListElementFields]]]
Making a LabelledGeneric[Some[Option[com.gu.fezziwig.ListElementFields]]]
Making a LabelledGeneric[Option[com.gu.fezziwig.ListElementFields]]
Making a LabelledGeneric[Some[com.gu.fezziwig.ListElementFields]]
Making a LabelledGeneric[com.gu.fezziwig.ListElementFields]
Making a LabelledGeneric[Option[scala.collection.Seq[com.gu.fezziwig.ListItem]]]
Making a LabelledGeneric[Some[scala.collection.Seq[com.gu.fezziwig.ListItem]]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.ListItem]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.ListItem]]
Making a LabelledGeneric[com.gu.fezziwig.ListItem]
Making a LabelledGeneric[Option[scala.collection.Seq[com.gu.fezziwig.BlockElement]]]
Making a LabelledGeneric[Some[scala.collection.Seq[com.gu.fezziwig.BlockElement]]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.BlockElement]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.BlockElement]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.BlockElement]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.BlockElement]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.BlockElement]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.BlockElement]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.ListItem]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.ListItem]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.ListItem]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.ListItem]]
Making a LabelledGeneric[com.gu.fezziwig.BlockElement]
Making a LabelledGeneric[Option[Option[com.gu.fezziwig.ListElementFields]]]
Making a LabelledGeneric[Some[Option[com.gu.fezziwig.ListElementFields]]]
Making a LabelledGeneric[Option[com.gu.fezziwig.ListElementFields]]
Making a LabelledGeneric[Some[com.gu.fezziwig.ListElementFields]]
Making a LabelledGeneric[com.gu.fezziwig.ListElementFields]
Making a LabelledGeneric[Option[scala.collection.Seq[com.gu.fezziwig.ListItem]]]
Making a LabelledGeneric[Some[scala.collection.Seq[com.gu.fezziwig.ListItem]]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.ListItem]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.ListItem]]
Making a LabelledGeneric[com.gu.fezziwig.ListItem]
Making a LabelledGeneric[Option[scala.collection.Seq[com.gu.fezziwig.BlockElement]]]
Making a LabelledGeneric[Some[scala.collection.Seq[com.gu.fezziwig.BlockElement]]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.BlockElement]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.BlockElement]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.BlockElement]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.BlockElement]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.BlockElement]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.BlockElement]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.ListItem]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.ListItem]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.ListItem]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.ListItem]]
Making a LabelledGeneric[com.gu.fezziwig.BlockElement]
Making a LabelledGeneric[Option[Option[com.gu.fezziwig.ListElementFields]]]
Making a LabelledGeneric[Some[Option[com.gu.fezziwig.ListElementFields]]]
Making a LabelledGeneric[Option[com.gu.fezziwig.ListElementFields]]
Making a LabelledGeneric[Some[com.gu.fezziwig.ListElementFields]]
Making a LabelledGeneric[com.gu.fezziwig.ListElementFields]
Making a LabelledGeneric[Option[scala.collection.Seq[com.gu.fezziwig.ListItem]]]
Making a LabelledGeneric[Some[scala.collection.Seq[com.gu.fezziwig.ListItem]]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.ListItem]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.ListItem]]
Making a LabelledGeneric[com.gu.fezziwig.ListItem]
Making a LabelledGeneric[Option[scala.collection.Seq[com.gu.fezziwig.BlockElement]]]
Making a LabelledGeneric[Some[scala.collection.Seq[com.gu.fezziwig.BlockElement]]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.BlockElement]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.BlockElement]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.BlockElement]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.BlockElement]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.BlockElement]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.BlockElement]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.ListItem]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.ListItem]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.ListItem]]
Making a LabelledGeneric[scala.collection.Seq[com.gu.fezziwig.ListItem]]
Making a LabelledGeneric[com.gu.fezziwig.StructB]
Making a LabelledGeneric[com.gu.fezziwig.StructC]
Making a LabelledGeneric[com.gu.fezziwig.StructD]
Making a LabelledGeneric[com.gu.fezziwig.StructB]
Making a LabelledGeneric[com.gu.fezziwig.StructC]
Making a LabelledGeneric[com.gu.fezziwig.StructD]
Making a LabelledGeneric[com.gu.fezziwig.StructB]
Making a LabelledGeneric[com.gu.fezziwig.StructC]
Making a LabelledGeneric[com.gu.fezziwig.StructD]
Making a LabelledGeneric[com.gu.fezziwig.StructB]
Making a LabelledGeneric[com.gu.fezziwig.StructC]
Making a LabelledGeneric[com.gu.fezziwig.StructD]
Making a LabelledGeneric[com.gu.fezziwig.StructA]
Making a LabelledGeneric[Option[Option[com.gu.fezziwig.StructC]]]
Making a LabelledGeneric[Some[Option[com.gu.fezziwig.StructC]]]
Making a LabelledGeneric[Option[com.gu.fezziwig.StructC]]
Making a LabelledGeneric[Some[com.gu.fezziwig.StructC]]
Making a LabelledGeneric[com.gu.fezziwig.StructC]
Making a LabelledGeneric[Option[scala.collection.Map[String,scala.collection.Seq[Int]]]]
Making a LabelledGeneric[Some[scala.collection.Map[String,scala.collection.Seq[Int]]]]
Making a LabelledGeneric[scala.collection.Map[String,scala.collection.Seq[Int]]]
Making a LabelledGeneric[scala.collection.Map[String,scala.collection.Seq[Int]]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Map[String,scala.collection.Seq[Int]]]
Making a LabelledGeneric[scala.collection.Map[String,scala.collection.Seq[Int]]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Map[String,scala.collection.Seq[Int]]]
Making a LabelledGeneric[scala.collection.Map[String,scala.collection.Seq[Int]]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[com.gu.fezziwig.Enum]
Making a LabelledGeneric[Option[Option[Int]]]
Making a LabelledGeneric[Some[Option[Int]]]
Making a LabelledGeneric[Option[Int]]
Making a LabelledGeneric[Some[Int]]
Making a LabelledGeneric[Option[Option[String]]]
Making a LabelledGeneric[Some[Option[String]]]
Making a LabelledGeneric[Option[String]]
Making a LabelledGeneric[Some[String]]
Making a LabelledGeneric[com.gu.fezziwig.StructB]
Making a LabelledGeneric[com.gu.fezziwig.StructD]
Making a LabelledGeneric[com.gu.fezziwig.StructA]
Making a LabelledGeneric[Option[Option[com.gu.fezziwig.StructC]]]
Making a LabelledGeneric[Some[Option[com.gu.fezziwig.StructC]]]
Making a LabelledGeneric[Option[com.gu.fezziwig.StructC]]
Making a LabelledGeneric[Some[com.gu.fezziwig.StructC]]
Making a LabelledGeneric[com.gu.fezziwig.StructC]
Making a LabelledGeneric[Option[scala.collection.Map[String,scala.collection.Seq[Int]]]]
Making a LabelledGeneric[Some[scala.collection.Map[String,scala.collection.Seq[Int]]]]
Making a LabelledGeneric[scala.collection.Map[String,scala.collection.Seq[Int]]]
Making a LabelledGeneric[scala.collection.Map[String,scala.collection.Seq[Int]]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Map[String,scala.collection.Seq[Int]]]
Making a LabelledGeneric[scala.collection.Map[String,scala.collection.Seq[Int]]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Map[String,scala.collection.Seq[Int]]]
Making a LabelledGeneric[scala.collection.Map[String,scala.collection.Seq[Int]]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[com.gu.fezziwig.Enum]
Making a LabelledGeneric[Option[Option[Int]]]
Making a LabelledGeneric[Some[Option[Int]]]
Making a LabelledGeneric[Option[Int]]
Making a LabelledGeneric[Some[Int]]
Making a LabelledGeneric[Option[Option[String]]]
Making a LabelledGeneric[Some[Option[String]]]
Making a LabelledGeneric[Option[String]]
Making a LabelledGeneric[Some[String]]
Making a LabelledGeneric[com.gu.fezziwig.StructB]
Making a LabelledGeneric[com.gu.fezziwig.StructD]
Making a LabelledGeneric[com.gu.fezziwig.StructA]
Making a LabelledGeneric[Option[Option[com.gu.fezziwig.StructC]]]
Making a LabelledGeneric[Some[Option[com.gu.fezziwig.StructC]]]
Making a LabelledGeneric[Option[com.gu.fezziwig.StructC]]
Making a LabelledGeneric[Some[com.gu.fezziwig.StructC]]
Making a LabelledGeneric[com.gu.fezziwig.StructC]
Making a LabelledGeneric[Option[scala.collection.Map[String,scala.collection.Seq[Int]]]]
Making a LabelledGeneric[Some[scala.collection.Map[String,scala.collection.Seq[Int]]]]
Making a LabelledGeneric[scala.collection.Map[String,scala.collection.Seq[Int]]]
Making a LabelledGeneric[scala.collection.Map[String,scala.collection.Seq[Int]]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Map[String,scala.collection.Seq[Int]]]
Making a LabelledGeneric[scala.collection.Map[String,scala.collection.Seq[Int]]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Map[String,scala.collection.Seq[Int]]]
Making a LabelledGeneric[scala.collection.Map[String,scala.collection.Seq[Int]]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[com.gu.fezziwig.Enum]
Making a LabelledGeneric[Option[Option[Int]]]
Making a LabelledGeneric[Some[Option[Int]]]
Making a LabelledGeneric[Option[Int]]
Making a LabelledGeneric[Some[Int]]
Making a LabelledGeneric[Option[Option[String]]]
Making a LabelledGeneric[Some[Option[String]]]
Making a LabelledGeneric[Option[String]]
Making a LabelledGeneric[Some[String]]
Making a LabelledGeneric[com.gu.fezziwig.StructB]
Making a LabelledGeneric[com.gu.fezziwig.StructD]
Making a LabelledGeneric[com.gu.fezziwig.StructA]
Making a LabelledGeneric[Option[Option[com.gu.fezziwig.StructC]]]
Making a LabelledGeneric[Some[Option[com.gu.fezziwig.StructC]]]
Making a LabelledGeneric[Option[com.gu.fezziwig.StructC]]
Making a LabelledGeneric[Some[com.gu.fezziwig.StructC]]
Making a LabelledGeneric[com.gu.fezziwig.StructC]
Making a LabelledGeneric[Option[scala.collection.Map[String,scala.collection.Seq[Int]]]]
Making a LabelledGeneric[Some[scala.collection.Map[String,scala.collection.Seq[Int]]]]
Making a LabelledGeneric[scala.collection.Map[String,scala.collection.Seq[Int]]]
Making a LabelledGeneric[scala.collection.Map[String,scala.collection.Seq[Int]]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Map[String,scala.collection.Seq[Int]]]
Making a LabelledGeneric[scala.collection.Map[String,scala.collection.Seq[Int]]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Map[String,scala.collection.Seq[Int]]]
Making a LabelledGeneric[scala.collection.Map[String,scala.collection.Seq[Int]]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[com.gu.fezziwig.Enum]
Making a LabelledGeneric[Option[Option[Int]]]
Making a LabelledGeneric[Some[Option[Int]]]
Making a LabelledGeneric[Option[Int]]
Making a LabelledGeneric[Some[Int]]
Making a LabelledGeneric[Option[Option[String]]]
Making a LabelledGeneric[Some[Option[String]]]
Making a LabelledGeneric[Option[String]]
Making a LabelledGeneric[Some[String]]
Making a LabelledGeneric[com.gu.fezziwig.StructB]
Making a LabelledGeneric[com.gu.fezziwig.StructD]
Making a LabelledGeneric[com.gu.fezziwig.StructWithOptional]
Making a LabelledGeneric[Option[Option[String]]]
Making a LabelledGeneric[Some[Option[String]]]
Making a LabelledGeneric[Option[String]]
Making a LabelledGeneric[Some[String]]
Making a LabelledGeneric[com.gu.fezziwig.StructWithOptional]
Making a LabelledGeneric[Option[Option[String]]]
Making a LabelledGeneric[Some[Option[String]]]
Making a LabelledGeneric[Option[String]]
Making a LabelledGeneric[Some[String]]
Making a LabelledGeneric[com.gu.fezziwig.StructA]
Making a LabelledGeneric[Option[Option[com.gu.fezziwig.StructC]]]
Making a LabelledGeneric[Some[Option[com.gu.fezziwig.StructC]]]
Making a LabelledGeneric[Option[com.gu.fezziwig.StructC]]
Making a LabelledGeneric[Some[com.gu.fezziwig.StructC]]
Making a LabelledGeneric[com.gu.fezziwig.StructC]
Making a LabelledGeneric[Option[scala.collection.Map[String,scala.collection.Seq[Int]]]]
Making a LabelledGeneric[Some[scala.collection.Map[String,scala.collection.Seq[Int]]]]
Making a LabelledGeneric[scala.collection.Map[String,scala.collection.Seq[Int]]]
Making a LabelledGeneric[scala.collection.Map[String,scala.collection.Seq[Int]]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Map[String,scala.collection.Seq[Int]]]
Making a LabelledGeneric[scala.collection.Map[String,scala.collection.Seq[Int]]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Map[String,scala.collection.Seq[Int]]]
Making a LabelledGeneric[scala.collection.Map[String,scala.collection.Seq[Int]]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[com.gu.fezziwig.Enum]
Making a LabelledGeneric[Option[Option[Int]]]
Making a LabelledGeneric[Some[Option[Int]]]
Making a LabelledGeneric[Option[Int]]
Making a LabelledGeneric[Some[Int]]
Making a LabelledGeneric[Option[Option[String]]]
Making a LabelledGeneric[Some[Option[String]]]
Making a LabelledGeneric[Option[String]]
Making a LabelledGeneric[Some[String]]
Making a LabelledGeneric[com.gu.fezziwig.StructB]
Making a LabelledGeneric[com.gu.fezziwig.StructD]
Making a LabelledGeneric[com.gu.fezziwig.StructA]
Making a LabelledGeneric[Option[Option[com.gu.fezziwig.StructC]]]
Making a LabelledGeneric[Some[Option[com.gu.fezziwig.StructC]]]
Making a LabelledGeneric[Option[com.gu.fezziwig.StructC]]
Making a LabelledGeneric[Some[com.gu.fezziwig.StructC]]
Making a LabelledGeneric[com.gu.fezziwig.StructC]
Making a LabelledGeneric[Option[scala.collection.Map[String,scala.collection.Seq[Int]]]]
Making a LabelledGeneric[Some[scala.collection.Map[String,scala.collection.Seq[Int]]]]
Making a LabelledGeneric[scala.collection.Map[String,scala.collection.Seq[Int]]]
Making a LabelledGeneric[scala.collection.Map[String,scala.collection.Seq[Int]]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Map[String,scala.collection.Seq[Int]]]
Making a LabelledGeneric[scala.collection.Map[String,scala.collection.Seq[Int]]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Map[String,scala.collection.Seq[Int]]]
Making a LabelledGeneric[scala.collection.Map[String,scala.collection.Seq[Int]]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[scala.collection.Seq[Int]]
Making a LabelledGeneric[com.gu.fezziwig.Enum]
Making a LabelledGeneric[Option[Option[Int]]]
Making a LabelledGeneric[Some[Option[Int]]]
Making a LabelledGeneric[Option[Int]]
Making a LabelledGeneric[Some[Int]]
Making a LabelledGeneric[Option[Option[String]]]
Making a LabelledGeneric[Some[Option[String]]]
Making a LabelledGeneric[Option[String]]
Making a LabelledGeneric[Some[String]]
Making a LabelledGeneric[com.gu.fezziwig.StructB]
Making a LabelledGeneric[com.gu.fezziwig.StructD]
Making a LabelledGeneric[com.gu.fezziwig.DefaultTestStruct]
Making a LabelledGeneric[Option[Int]]
Making a LabelledGeneric[Some[Int]]
Making a LabelledGeneric[Option[Option[Int]]]
Making a LabelledGeneric[Some[Option[Int]]]
Making a LabelledGeneric[com.gu.fezziwig.DefaultTestStruct]
Making a LabelledGeneric[Option[Int]]
Making a LabelledGeneric[Some[Int]]
Making a LabelledGeneric[Option[Option[Int]]]
Making a LabelledGeneric[Some[Option[Int]]]
Making a LabelledGeneric[com.gu.fezziwig.DefaultTestStruct]
Making a LabelledGeneric[Option[Int]]
Making a LabelledGeneric[Some[Int]]
Making a LabelledGeneric[Option[Option[Int]]]
Making a LabelledGeneric[Some[Option[Int]]]
Making a LabelledGeneric[com.gu.fezziwig.DefaultTestStruct]
Making a LabelledGeneric[Option[Int]]
Making a LabelledGeneric[Some[Int]]
Making a LabelledGeneric[Option[Option[Int]]]
Making a LabelledGeneric[Some[Option[Int]]]
Making a LabelledGeneric[com.gu.fezziwig.DefaultTestStruct]
Making a LabelledGeneric[Option[Int]]
Making a LabelledGeneric[Some[Int]]
Making a LabelledGeneric[Option[Option[Int]]]
Making a LabelledGeneric[Some[Option[Int]]]
Making a LabelledGeneric[com.gu.fezziwig.DefaultTestStruct]
Making a LabelledGeneric[Option[Int]]
Making a LabelledGeneric[Some[Int]]
Making a LabelledGeneric[Option[Option[Int]]]
Making a LabelledGeneric[Some[Option[Int]]]
Making a LabelledGeneric[com.gu.fezziwig.DefaultTestStruct]
Making a LabelledGeneric[Option[Int]]
Making a LabelledGeneric[Some[Int]]
Making a LabelledGeneric[Option[Option[Int]]]
Making a LabelledGeneric[Some[Option[Int]]]
Making a LabelledGeneric[com.gu.fezziwig.DefaultTestStruct]
Making a LabelledGeneric[Option[Int]]
Making a LabelledGeneric[Some[Int]]
Making a LabelledGeneric[Option[Option[Int]]]
Making a LabelledGeneric[Some[Option[Int]]]
Making a LabelledGeneric[com.gu.fezziwig.DefaultTestStruct]
Making a LabelledGeneric[Option[Int]]
Making a LabelledGeneric[Some[Int]]
Making a LabelledGeneric[Option[Option[Int]]]
Making a LabelledGeneric[Some[Option[Int]]]
Making a LabelledGeneric[com.gu.fezziwig.DefaultTestStruct]
Making a LabelledGeneric[Option[Int]]
Making a LabelledGeneric[Some[Int]]
Making a LabelledGeneric[Option[Option[Int]]]
Making a LabelledGeneric[Some[Option[Int]]]
Making a LabelledGeneric[com.gu.fezziwig.DefaultTestStruct]
Making a LabelledGeneric[Option[Int]]
Making a LabelledGeneric[Some[Int]]
Making a LabelledGeneric[Option[Option[Int]]]
Making a LabelledGeneric[Some[Option[Int]]]
Making a LabelledGeneric[com.gu.fezziwig.DefaultTestStruct]
Making a LabelledGeneric[Option[Int]]
Making a LabelledGeneric[Some[Int]]
Making a LabelledGeneric[Option[Option[Int]]]
Making a LabelledGeneric[Some[Option[Int]]]
```
</details>

davidfurey commented:
I'm not really sure how cacheImplicits could be used within a macro context. 

I'm starting to question whether fully-automatic is actually a good approach. Content-api-models creates named decoders for many, but not all, of the structs. I suspect that this is because fully auto was too much for the compiler. 

Perhaps instead of spending more time on fully auto, we should instead provide better error messages for the semi-auto approach? At the moment, if you are missing an Decoder dependency of Decoder[A] it complains about not being able to find Decoder[A] but it doesn't tell you that this is because Decoder[B] is missing. If we could fix this, it would be much easier to migrate to semi-auto.

emdash-ie commented:
> I'm not really sure how cacheImplicits could be used within a macro context.

Yeah, I had a look but can’t figure out anything useful.

> I'm starting to question whether fully-automatic is actually a good approach. Content-api-models creates named decoders for many, but not all, of the structs. I suspect that this is because fully auto was too much for the compiler.
> 
> Perhaps instead of spending more time on fully auto, we should instead provide better error messages for the semi-auto approach? At the moment, if you are missing an Decoder dependency of Decoder[A] it complains about not being able to find Decoder[A] but it doesn't tell you that this is because Decoder[B] is missing. If we could fix this, it would be much easier to migrate to semi-auto.

Ok, that’s tempting! I will take a look when I can and see if I can work anything out.

emdash-ie commented:
Found someone else pursuing this LabelledGeneric approach [back in 2015](https://github.com/twitter/scrooge/issues/170)! 

Leads to [this example from Travis Brown](https://github.com/travisbrown/scrooge-circe-demo/blob/master/src/main/scala/demo/package.scala) which seems simpler than what I’ve done, might be worth stealing from that.

emdash-ie committed:
Revert "Remove necessity for manual semi-auto derivation"

Fully auto derivation like this performs really badly! For reasons I
don’t understand (probably related to them being whitebox macros?), the
compiler ends up calling the macro for many types that are not thrift
structs, e.g. Option[Option[Int]] – I think for every type that appears
in the generic representation.

I can’t work out how to fix it yet, so I’m removing it for now.

This reverts commit 37f39c11d450dc550fe21d6959a20612fdbb0c70.

emdash-ie committed:
Update comment (macro imports)

emdash-ie commented:
I’ve updated content-api-models to work with this PR: https://github.com/guardian/content-api-models/pull/239

The semi-auto approach is a little tedious, but it’s not _too_ bad – I’m tempted to accept it.

emdash-ie committed:
Alter CNil case to remove warning

Without this type signature, the compiler throws no warning (whereas
with it, or with the case removed, we get a warning). I believe this
case still can’t happen, but I think avoiding the warning and throwing
an exception is probably best.

emdash-ie committed:
Encode and decode UnknownUnionField values

The classes scrooge generates for thrift unions include
`UnknownUnionField` values, which are for writing code that doesn’t
crash when new fields/variants are added.

The existing behaviour in fezziwig for these values is that the encoder
successfully encodes them (as null), but the decoder throws an error
when decoding null (unless the field is also marked optional).

Rather than mimic the existing behaviour (which I also don’t know how to
do), I’ve chosen to preserve the field metadata (names, ids, and types)
but drop the field values. This seems like a sensible way to handle
unknown union fields for now, as preserving the values would require a
bit more effort. (I guess we’d need to convert the raw bytes to/from
circe’s Json type and store that? Hmmm…)

emdash-ie committed:
Use nicer shapeless syntax for TField implicit

This nicer syntax avoids the need for witness vals, and is much less
verbose.

It would potentially be worth using this syntax in the macros as well,
but I imagine it’ll be a little complicated to get it right and I’m not
sure it’s worth it.

emdash-ie committed:
Rename ...Generic -> ...LabelledGeneric

While generating arbitrary values from generics for a 10% project, I
needed to make versions of these macros for shapeless’ non-labelled
Generic class, so I’d rather reserve those names in case we want them in
future.

JustinPinner reviewed (approved):
There have been better brains than mine engaged on this, but I _think_ I know enough about the reasoning and approach taken to be able to approve. And it's a library so it won't immediately make things catch fire when released. And it'll be a major version increment, right? So with that said: 👍

emdash-ie commented:
> And it's a library so it won't immediately make things catch fire when released. And it'll be a major version increment, right?

Right and right!

emdash-ie commented:
I had a look at the effect of these changes on compilation times, by comparing compile times on [content-api-models](https://github.com/guardian/content-api-models) before ([commit d77f319f](https://github.com/guardian/content-api-models/commit/d77f319f)) and after ([commit d5bbf4646](https://github.com/guardian/content-api-models/commit/d5bbf4646568f8f109d07e3bfb78ff3ca03342d6)) the change. (I produced the new snapshot version by running `sbt +publishLocal` in the fezziwig repository.)

I made a small number of measurements and found a decrease of ~10% in compilation times. More measurements would give more confidence this can’t be random variation, but I did vary the order I took the measurements in to try and make sure it wasn’t down to CPU warmup or etc.

## `sbt compile`

I ran `sbt clean > /dev/null; time sbt compile > /dev/null`, and got the following outputs:

- before: 
  - 123.31s
  - 121.31s
- after: 
  - 108.10s
  - 108.83s

So we have a speedup of 11% in compilation! (At the cost of having to semi-automatically derive the encoders and decoders)

## `sbt +compile`

I also tried including cross-compilation to see if that would make a difference: for these measurements I ran `sbt clean > /dev/null; time sbt +compile > /dev/null`.

- before:
  - 219.46s
- after:
  - 196.98s

This shows a speedup of 10%.

## `sbt +test`

For these measurements I ran the tests to see what difference that made: `sbt clean; time sbt "project content-api-models-json; +test" > /dev/null`.

- before:
  - 242.57s
  - 231.14s
- after:
  - 204.71s
  - 195.15s

Speedup here is 15%.
